embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java

```diff
@@ -0,0 +1,608 @@
+            }
+        }
+        return patIndex;
+    }
+
+    private static boolean isValidRange(long v, int lower, int upper)
+    {
+        return lower <= v && v <= upper;
+    }
+
+    private static boolean isSpace(char c)
+    {
+        return c == ' ' || c == '\t' || c == '\n' ||
+                c == '\u000b' || c == '\f' || c == '\r';
+    }
+
+    static boolean isDigit(char c)
+    {
+        return '0' <= c && c <= '9';
+    }
+
+    static boolean isSign(char c)
+    {
+        return c == '+' || c == '-';
+    }
+
+    static int toInt(char c)
+    {
+        return c - '0';
+    }
+
+    private static Set<Format> numPatterns; // CDdeFGgHIjkLlMmNQRrSsTUuVvWwXxYy
+    static {
+        numPatterns = new HashSet<>();
+        numPatterns.add(Format.FORMAT_CENTURY); // 'C'
+        // D
+        numPatterns.add(Format.FORMAT_DAY); // 'd'
+        numPatterns.add(Format.FORMAT_DAY_S); // 'e'
+        // F
+        numPatterns.add(Format.FORMAT_WEEKYEAR); // 'G'
+        numPatterns.add(Format.FORMAT_WEEKYEAR_SHORT); // 'g'
+        numPatterns.add(Format.FORMAT_HOUR); // 'H'
+        numPatterns.add(Format.FORMAT_HOUR_M); // 'I'
+        numPatterns.add(Format.FORMAT_DAY_YEAR); // 'j'
+        numPatterns.add(Format.FORMAT_HOUR_BLANK); // 'k'
+        numPatterns.add(Format.FORMAT_MILLISEC); // 'L'
+        numPatterns.add(Format.FORMAT_HOUR_S); // 'l'
+        numPatterns.add(Format.FORMAT_MINUTES); // 'M'
+        numPatterns.add(Format.FORMAT_MONTH); // 'm'
+        numPatterns.add(Format.FORMAT_NANOSEC); // 'N'
+        // Q, R, r
+        numPatterns.add(Format.FORMAT_SECONDS); // 'S'
+        numPatterns.add(Format.FORMAT_EPOCH); // 's'
+        // T
+        numPatterns.add(Format.FORMAT_WEEK_YEAR_S); // 'U'
+        numPatterns.add(Format.FORMAT_DAY_WEEK2); // 'u'
+        numPatterns.add(Format.FORMAT_WEEK_WEEKYEAR); // 'V'
+        // v
+        numPatterns.add(Format.FORMAT_WEEK_YEAR_M); // 'W'
+        numPatterns.add(Format.FORMAT_DAY_WEEK); // 'w'
+        // X, x
+        numPatterns.add(Format.FORMAT_YEAR_LONG); // 'Y'
+        numPatterns.add(Format.FORMAT_YEAR_SHORT); // 'y'
+    }
+
+    private static boolean matchAtNumPatterns(Token token) // NUM_PATTERN_P
+    {
+        if (token == null) {
+            return false;
+        }
+
+        // TODO
+        Format f = token.getFormat();
+        if (f == Format.FORMAT_STRING && isDigit(((String)token.getData()).charAt(0))) {
+            return true;
+
+        } else if (numPatterns.contains(f)) {
+            return true;
+
+        }
+        return false;
+    }
+
+    private static RuntimeException newInvalidDataException() // Token token, int v
+    {
+        return new RuntimeException("Invalid Data"); // TODO InvalidDataException
+    }
+
+    private final ThreadContext context;
+    // Use RubyDateFormatter temporarily because it has useful lexer, token and format types
+    private final RubyDateFormatter dateFormat;
```

Comment:
Yes. And please try to make RubyDateParser#parseInternal and RubyDateFormatter#format independent from Ruby runtime (org.jruby.Ruby). If they don't need Ruby runtime, we can simplify both embulk-core and plugin code because org.embulk.spi.time.TimestampParser/TimestampFormatter don't need ScriptingContainer.